### PR TITLE
Dataes 199 allowing generic result mapping

### PR DESCRIPTION
**Pull request by dandonovsc**
A new search result mapper that is bound to input and output generic types, allowing the user to implement the interface with their own domain object to be mapped from the index source data

### CODE REVIEW

File: src/test/java/org/springframework/data/elasticsearch/core/ElasticsearchTemplateTests.java [hunk 1]
```diff
@@ -1903,6 +1901,47 @@ public void shouldComposeObjectsReturnedFromHeterogeneousIndexes() {
...
+	 * This is a demonstration to show composing entities out of index that are different type than the index class.
+	 */
+	public void shouldComposeIntermittentObjectFromResult() {
+
+        String indexName = Person.class.getAnnotation(Document.class).indexName();
+        Person person = new Person();
+        person.setName("John Smith");
+        person.setCar(Lists.newArrayList(new Car("BMW", "330i")));
+
+		IndexQuery idxQuery1 = new IndexQueryBuilder().withIndexName(indexName).withObject(person).build();
```
Comment (martin-g) [inline src/test/java/org/springframework/data/elasticsearch/core/ElasticsearchTemplateTests.java]: wrong whitespaces/indentation

**Comment by dandonovsc** [comment]
I have pushed the updated code with the correct formatting

**Comment by dandonovsc** [comment]
@martin-g , what is the status of this pull request?

**Comment by martin-g** [comment]
I'm not maintaiber of this project.
I'm a user like you. Just interested in the changes, that's why I review
the PRs and comment on them when I see something suboptimal.

**Comment by xhaggi** [comment]
Already done in master.

**Pull request closed without merging**